### PR TITLE
Hashlookup

### PR DESCRIPTION
**Pull request by gallypette**
Hi,

Here the PR.
It has stalled as I would like to paginate hashlookup results and that is not implemented in that end yet.

### CODE REVIEW

**Comment by adulau** [comment]
@gallypette Thanks for the reminder. It's now implemented https://github.com/adulau/hashlookup-server/commit/baf2b82f7b961194ee361abf2c33d294ed86d71f and exposed on the public hashlookup instance.

File: plugins/src/hashlookup/convert.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+package main
+
+import (
+	"errors"
+	"github.com/blastrain/vitess-sqlparser/sqlparser"
+)
+
+/*
+ * Convert SQL statement to the object expected by the data source
```
Comment (aleksejsv) [inline plugins/src/hashlookup/convert.go]: * Convert SQL statement to the list of [field,value]

File: plugins/src/hashlookup/hashlookup.go [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+		for _, types := range relation.From.VarTypes {
+			types.RegexCompiled = regexp.MustCompile(types.Regex)
+		}
+
+		for _, types := range relation.To.VarTypes {
+			types.RegexCompiled = regexp.MustCompile(types.Regex)
+		}
+	}
+
+	fmt.Printf("HTTP %s: %#v\n\n", source.Name, p)
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: // fmt.Printf("Hashlookup %s: %#v\n\n", source.Name, p)

File: plugins/src/hashlookup/hashlookup.go [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+	return nil
+}
+
+func (p *plugin) Search(stmt *sqlparser.Select) ([]map[string]interface{}, map[string]interface{}, error) {
+
+	// Storage for the results to return
+	results := []map[string]interface{}{}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: // Convert SQL statement (comment before the next function)

File: plugins/src/hashlookup/hashlookup.go [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+	var entries []map[string]interface{}
+
+	for _, body := range unpacked {
+		// Struct to store statistics data
+		// when the amount of returned entries is too large
+		stats := pdk.NewStats()
+
+		for _, field := range p.source.StatsFields {
+			stats.Fields[field] = sortedmap.New(10, desc.Int)
+		}
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: Stats info does nothing here

File: plugins/src/hashlookup/hashlookup.go [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+		parents := new(gabs.Container)
+
+		// TODO fetch children / parents according to the LIMIT
+		// if the children/parent key exist and the count hits hashlookup limits,
+		// we query hashlookup again up the the query' LIMIT
+
+		if jsonParsed.Exists("children") {
+			if err != nil {
+				return nil, nil, err
+			}
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: "err" of what is here?

File: plugins/src/hashlookup/hashlookup.go [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+		if jsonParsed.Exists("parents") {
+			// Here we run the new query
+			gabsParents, err := p.getRelated("parents", jsonParsed.S("SHA-1").Data().(string), p.limit)
+			if err != nil {
+				return nil, nil, err
+			}
+			parents = gabsParents.Path("parents")
+			if err != nil {
+				return nil, nil, err
+			}
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: "err" of what is here?

File: plugins/src/hashlookup/hashlookup.go [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+	for _, field := range p.source.StatsFields {
+		stats.Fields[field] = sortedmap.New(10, desc.Int)
+	}
+
+	for _, entry := range entries {
+
+		// Stop when results count is over the limit
+		if counter >= p.limit {
+			// Uncomment in real plugin
+			//cancel()
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: "cancel" is nil at this point

File: plugins/src/hashlookup/hashlookup.go [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+	 * STEP 8.
+	 *
+	 * Stop the plugin when main service stops,
+	 * drop all connections correctly
+	 */
+
+	// ctx, cancel := context.WithTimeout(context.Background(), p.source.Timeout)
+	// defer cancel()
+
+	// return p.client.Disconnect(ctx)
```
Comment (aleksejsv) [inline plugins/src/hashlookup/hashlookup.go]: Code left from a template. Can be replaced with just `No error to check, so return nil`

File: plugins/src/hashlookup/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# Hashlookup plugin
+
+Plugin to query [hashlookup services](https://github.com/hashlookup).
+For instance hashlookup.circl.lu
+
+Sample command to use plugin:
+```sh
+curl 'https://localhost:443/api?uuid=auth-key&sql=FROM+hashlookup+WHERE+sha1=%27deac5aeda66017c25a9e21f36f5ee618d2ad9d3d%27'
+```
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/README.md]: How to use bulk requests? Maybe other variants?

File: plugins/src/hashlookup/select.go [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+		// TODO: more precise
+		if _, ok := expr.Expr.(*sqlparser.FuncExpr); ok {
+			return true
+		}
+	}
+
+	return false
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go:28]: /*
 * Handle single "field operator value" expression.
 *
 * Receives:
 *     expr     - SQL expression to process
 *     topLevel - whether it's a top level expression
 *     parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+	switch comparisonExpr.Operator {
+	case "=":
+		fields = append(fields, [2]string{colNameStr, fmt.Sprintf("%s", rightIntf)})
+	default:
+		return nil, errors.New("'=' operator is supported only")
+	}
+
+	return fields, nil
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go:55]: /*
 * Handle "expression AND expression".
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+	rightStr, err := handleSelectWhere(&rightExpr, false, expr)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := append(leftStr, rightStr...)
+	if topLevel {
+		fields = append(fields, [2]string{"operator", "and"})
+	}
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: Is it useful to add "operator" parameter for the hashlookup service?

File: plugins/src/hashlookup/select.go [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+	return fields, nil
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go:78]: /*
 * Handle "expression OR expression".
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+	rightStr, err := handleSelectWhere(&rightExpr, false, expr)
+	if err != nil {
+		return nil, err
+	}
+
+	fields := append(leftStr, rightStr...)
+	if topLevel {
+		fields = append(fields, [2]string{"operator", "or"})
+	}
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: Is it useful to add "operator" parameter for the hashlookup service?

File: plugins/src/hashlookup/select.go [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+
+	return fields, nil
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go:101]: Is it useful to add "BETWEEN" statement for the hashlookup service?

File: plugins/src/hashlookup/select.go [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	// Build resulting fields list
+	var fields [][2]string
+	fields = append(fields, [2]string{colNameStr + "_from", from})
+	fields = append(fields, [2]string{colNameStr + "_to", to})
+
+	return fields, nil
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go:153]: /*
 * Handle top level or groups of expressions.
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+		isThisTopLevel = true
+	}
+
+	return handleSelectWhere(&boolExpr, isThisTopLevel, parent)
+}
+
+func buildNestedFuncStrValue(nestedFunc *sqlparser.FuncExpr) (string, error) {
+	return "", errors.New("Unsupported function: " + nestedFunc.Name.String())
+}
+
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go:170]: /*
 * Check the right part of the expression
 * and return its value of specific type.
 *
 * Receives SQL expression to process
 */

File: plugins/src/hashlookup/select.go [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+	case *sqlparser.OrExpr:
+		return handleSelectWhereOrExpr(expr, topLevel, parent)
+
+	case *sqlparser.IsExpr:
+		return nil, errors.New("'is' expression currently not supported")
+
+	case *sqlparser.NotExpr:
+		return nil, errors.New("'not' expression currently not supported")
+
+	case *sqlparser.RangeCond:
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: Is it useful to add "BETWEEN" statement for the hashlookup service?

File: plugins/src/hashlookup/select.go [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+		// TODO: more precise
+		if _, ok := expr.Expr.(*sqlparser.FuncExpr); ok {
+			return true
+		}
+	}
+
+	return false
+}
+
+func handleSelectWhereComparisonExpr(expr *sqlparser.Expr, topLevel bool, parent *sqlparser.Expr) ([][2]string, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: /*
 * Handle single "field operator value" expression.
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+	case "=":
+		fields = append(fields, [2]string{colNameStr, fmt.Sprintf("%s", rightIntf)})
+	default:
+		return nil, errors.New("'=' operator is supported only")
+	}
+
+	return fields, nil
+}
+
+func handleSelectWhereAndExpr(expr *sqlparser.Expr, topLevel bool, parent *sqlparser.Expr) ([][2]string, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: /*
 * Handle "expression AND expression".
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+
+	fields := append(leftStr, rightStr...)
+	if topLevel {
+		fields = append(fields, [2]string{"operator", "and"})
+	}
+
+	return fields, nil
+}
+
+func handleSelectWhereOrExpr(expr *sqlparser.Expr, topLevel bool, parent *sqlparser.Expr) ([][2]string, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: /*
 * Handle "expression OR expression".
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+	if topLevel {
+		fields = append(fields, [2]string{"operator", "or"})
+	}
+
+	return fields, nil
+}
+
+// Between a and b.
+// The meaning is equal to range query
+func handleSelectWhereBetweenExpr(expr *sqlparser.Expr, topLevel bool) ([][2]string, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: Is it useful to add "BETWEEN" statement for the hashlookup service?

File: plugins/src/hashlookup/select.go [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+
+	// Build resulting fields list
+	var fields [][2]string
+	fields = append(fields, [2]string{colNameStr + "_from", from})
+	fields = append(fields, [2]string{colNameStr + "_to", to})
+
+	return fields, nil
+}
+
+func handleSelectWhereParenExpr(expr *sqlparser.Expr, topLevel bool, parent *sqlparser.Expr) ([][2]string, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: /*
 * Handle top level or groups of expressions.
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

File: plugins/src/hashlookup/select.go [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	return handleSelectWhere(&boolExpr, isThisTopLevel, parent)
+}
+
+func buildNestedFuncStrValue(nestedFunc *sqlparser.FuncExpr) (string, error) {
+	return "", errors.New("Unsupported function: " + nestedFunc.Name.String())
+}
+
+func buildComparisonExprRightStr(expr sqlparser.Expr) (interface{}, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: /*
 * Check the right part of the expression
 * and return its value of specific type.
 *
 * Receives SQL expression to process
 */

File: plugins/src/hashlookup/select.go [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+		rightStr = sqlparser.String(expr)
+
+	default:
+		return nil, fmt.Errorf("Unexpected SQL expression right part's type: %T", expr)
+	}
+
+	return rightStr, err
+}
+
+func handleSelectWhere(expr *sqlparser.Expr, topLevel bool, parent *sqlparser.Expr) ([][2]string, error) {
```
Comment (aleksejsv) [inline plugins/src/hashlookup/select.go]: /*
 * Handle WHERE statement.
 *
 * Receives:
 * expr     - SQL expression to process
 * topLevel - whether it's a top level expression
 * parent   - container of the expression
 */

**Comment by aleksejsv** [comment]
@gallypette sorry, during repo cleaning this PR was automatically closed somehow and can't be reopened. Please create a new one. Also any updates on Hashlookup plugin?

**Comment by adulau** [comment]
It's kind of complex to create new modules in graphoscope, having a generic [misp-modules](https://github.com/MISP/misp-modules) API interface.  @gallypette what do you think?

**Pull request closed without merging**